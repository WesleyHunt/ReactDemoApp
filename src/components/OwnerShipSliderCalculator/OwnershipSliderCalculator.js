import React, { useState, useEffect } from 'react';
import currency from 'currency.js';
import styles, { dot } from './ownershipSliderCalculator.module.css'
import Slider from '@material-ui/core/Slider'
import { withStyles } from '@material-ui/styles';

const ColoredSlider = withStyles({
  root: {
    width: '90%',
    alignSelf: 'center'
  },
  thumb: {
    color: 'black',
    backgroundColor: 'white',
    border: '1px solid currentColor',
    height: 25,
    width: 25,
    bottom: -2
  },
  track: {
    color: '#00CFCC',
    height: 10,
    borderRadius: 5
  },
  rail: {
    color: '#b9b9b9',
    height: 10,
    borderRadius: 5
  }
})(Slider);

export const calculateOwnership = (initialInvestment, monthlyInvestment, yearsRenting) => {
  return (parseInt(initialInvestment, 10) + parseInt(monthlyInvestment, 10) * parseInt(yearsRenting, 10) * 12) ?? "error"
}

export const formatToCurrency = (value) => {
  return currency(value, { precision: '0' }).format()
}


const OwnershipSliderCalculator = ({ rent, initialMin, initialMax, monthlyMax, yearsMax }) => {
  const [initialInvestment, setInitialInvestment] = useState(0)
  const [monthlyInvestment, setMonthlyInvestment] = useState(0)
  const [yearsWithCompany, setYearsWithCompany] = useState(1)
  const [ownershipValue, setOwnershipValue] = useState(0)
  const [yearsLabel, setYearsLabel] = useState("year")

  useEffect(() => {
    setOwnershipValue(calculateOwnership(initialInvestment, monthlyInvestment, yearsWithCompany))
  });

  return (

    <div className={styles.container}>

      <div className={styles.greyBoldText + " " + styles.leftHeaderMarginStyling}>MONTHLY PAYMENT</div>
      <section className={styles.row + " " + styles.leftHeaderMarginStyling}>
        <div className={styles.blackText}>${formatToCurrency(rent)} +</div>
        <div className={styles.blueText} style={{ marginLeft: "5px" }}>${formatToCurrency(initialInvestment)} investment</div>
      </section>
      <div className={styles.lineDivider} />

      <div className={styles.outerContainer}>
        <div className={styles.dotContainer}>
          <dot />
          <dot />
          <dot />
          <dot />
        </div>

        <div className={styles.innerContainer}>

          <section className={styles.row}>
            <div className={styles.greyBoldText + ' ' + styles.strikeThrough + ' ' + styles.leftMarginStyling}>SECURITY DEPOSIT</div>
            <div className={styles.greyBoldText}> &#x2192; INITIAL INVESTMENT</div>
          </section>
          <div className={styles.blackText} style={{ alignSelf: 'center' }}>${formatToCurrency(initialInvestment)}</div>
          <ColoredSlider
            className={styles.slider}
            min={initialMin ?? 500}
            max={initialMax ?? 5000}
            step={25}
            onChange={(object, value) =>
              setInitialInvestment(value)
            }
          />

          <div className={styles.greyBoldText + ' ' + styles.leftTopMarginStyling}>MONTHLY INVESTMENT</div>
          <div className={styles.blackText} style={{ alignSelf: 'center' }}>${formatToCurrency(monthlyInvestment)}</div>
          <ColoredSlider
            className={styles.slider}
            min={0}
            max={monthlyMax ?? 400}
            step={5}
            onChange={(object, value) =>
              setMonthlyInvestment(value)
            }
          />

          <div className={styles.greyBoldText + ' ' + styles.leftTopMarginStyling}>YEARS WITH UP&UP</div>
          <div className={styles.blackText} style={{ alignSelf: 'center' }}>{yearsWithCompany} {yearsLabel}</div>
          <ColoredSlider
            className={styles.slider}
            min={1}
            max={yearsMax ?? 5}
            step={1}
            onChange={(object, value) => {
              setYearsWithCompany(value)
              if (value > 1) {
                setYearsLabel("years")
              } else {
                setYearsLabel("year")
              }
            }
            }
          />

          <div className={styles.greyBoldText + ' ' + styles.leftTopMarginStyling}>OWNERSHIP VALUE</div>
          <div className={styles.blueTextLarge} style={{ alignSelf: 'center', marginTop: "10px", marginBottom: "7px" }}>${formatToCurrency(ownershipValue)}</div>
          <div className={styles.blackSmallText + ' ' + styles.centerTextStyling}>
            Use your ${formatToCurrency(ownershipValue)} ownership stake as a down payment on any home, transfer it to your next up&up rental, or cash out.</div>

        </div>
      </div>
    </div>
  )
}

export default OwnershipSliderCalculator;