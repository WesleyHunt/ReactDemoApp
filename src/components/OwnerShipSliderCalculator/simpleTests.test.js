import React from 'react';
import renderer from 'react-test-renderer';
import OwnershipSliderCalculator, {calculateOwnership} from './OwnershipSliderCalculator'


test('Component UI snapshot with props', () => {
    const component = renderer
    .create(<OwnershipSliderCalculator rent={2400} initialMin={1200} initialMax={4000} monthlyMax={800} yearsMax={5}/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

});


test('OwnershipCalculations do not fail/crash with various inputs', () => {
    expect(calculateOwnership(null, null, null)).toStrictEqual(expect.anything())
    expect(calculateOwnership(null, 100, null)).toStrictEqual(expect.anything())
    expect(calculateOwnership(null, "helloworld", null)).toStrictEqual(expect.anything())
    expect(calculateOwnership(1,2,3)).toStrictEqual(expect.anything())
    expect(calculateOwnership('1', '2', '3')).toStrictEqual(expect.anything())
    expect(calculateOwnership("hellow", "helloworld","world")).toStrictEqual(expect.anything())
});

    /**redundant check: function calculateOwnership (initialInvestment, monthlyInvestment, monthsRenting) => {
    return initialInvestment + (monthlyInvestment * monthsRenting)
    return 1500 + (100 * yearsRenting*12)
    return 1500 + (100 * 1*12) = 2700
    **/
test('OwnershipCalculations produces correct value', () => {
    expect(calculateOwnership(1500,100,1)).toStrictEqual(2700)
});



