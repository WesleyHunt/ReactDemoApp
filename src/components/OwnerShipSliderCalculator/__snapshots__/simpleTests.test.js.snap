// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component UI snapshot 1`] = `
<div
  className="container"
>
  <div
    className="greyBoldText leftMarginStyling"
  >
    MONTHLY PAYMENT
  </div>
  <section
    className="row leftMarginStyling"
  >
    <div
      className="blackText"
    >
      $
      0
       +
    </div>
    <div
      className="blueText"
      style={
        Object {
          "marginLeft": "5px",
        }
      }
    >
      $
      0
       investment
    </div>
  </section>
  <div
    className="lineDivider"
  />
  <section
    className="row"
  >
    <div
      className="greyBoldText strikeThrough leftMarginStyling"
    >
      SECURITY DEPOSIT
    </div>
    <div
      className="greyBoldText"
    >
       → INITIAL INVESTMENT
    </div>
  </section>
  <div
    className="blackText"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    $
    0
  </div>
  <span
    className="MuiSlider-root WithStyles(ForwardRef(Slider))-root-1 MuiSlider-colorPrimary slider"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-4"
    />
    <span
      className="MuiSlider-track WithStyles(ForwardRef(Slider))-track-3"
      style={
        Object {
          "left": "0%",
          "width": "NaN%",
        }
      }
    />
    <input
      type="hidden"
      value="NaN"
    />
    <span
      aria-orientation="horizontal"
      aria-valuemax={5000}
      aria-valuemin={500}
      aria-valuenow={NaN}
      className="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-2 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "NaN%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="greyBoldText leftTopMarginStyling"
  >
    MONTHLY INVESTMENT
  </div>
  <div
    className="blackText"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    $
    0
  </div>
  <span
    className="MuiSlider-root WithStyles(ForwardRef(Slider))-root-1 MuiSlider-colorPrimary slider"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-4"
    />
    <span
      className="MuiSlider-track WithStyles(ForwardRef(Slider))-track-3"
      style={
        Object {
          "left": "0%",
          "width": "NaN%",
        }
      }
    />
    <input
      type="hidden"
      value="NaN"
    />
    <span
      aria-orientation="horizontal"
      aria-valuemax={400}
      aria-valuemin={0}
      aria-valuenow={NaN}
      className="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-2 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "NaN%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="greyBoldText leftTopMarginStyling"
  >
    YEARS WITH UP&UP
  </div>
  <div
    className="blackText"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    1
     
    year
  </div>
  <span
    className="MuiSlider-root WithStyles(ForwardRef(Slider))-root-1 MuiSlider-colorPrimary slider"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-4"
    />
    <span
      className="MuiSlider-track WithStyles(ForwardRef(Slider))-track-3"
      style={
        Object {
          "left": "0%",
          "width": "NaN%",
        }
      }
    />
    <input
      type="hidden"
      value="NaN"
    />
    <span
      aria-orientation="horizontal"
      aria-valuemax={5}
      aria-valuemin={1}
      aria-valuenow={NaN}
      className="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-2 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "NaN%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="greyBoldText leftTopMarginStyling"
  >
    OWNERSHIP VALUE
  </div>
  <div
    className="blueTextLarge"
    style={
      Object {
        "alignSelf": "center",
        "marginBottom": "7px",
        "marginTop": "10px",
      }
    }
  >
    $
    0
  </div>
  <div
    className="greyText centerTextStyling"
  >
    Use your $
    0
     ownership stake as a down payment on any home, transfer it to your next up&up rental, or cash out.
  </div>
</div>
`;

exports[`Component UI snapshot with props 1`] = `
<div
  className="container"
>
  <div
    className="greyBoldText leftMarginStyling"
  >
    MONTHLY PAYMENT
  </div>
  <section
    className="row leftMarginStyling"
  >
    <div
      className="blackText"
    >
      $
      2,400
       +
    </div>
    <div
      className="blueText"
      style={
        Object {
          "marginLeft": "5px",
        }
      }
    >
      $
      0
       investment
    </div>
  </section>
  <div
    className="lineDivider"
  />
  <section
    className="row"
  >
    <div
      className="greyBoldText strikeThrough leftMarginStyling"
    >
      SECURITY DEPOSIT
    </div>
    <div
      className="greyBoldText"
    >
       → INITIAL INVESTMENT
    </div>
  </section>
  <div
    className="blackText"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    $
    0
  </div>
  <span
    className="MuiSlider-root WithStyles(ForwardRef(Slider))-root-1 MuiSlider-colorPrimary slider"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-4"
    />
    <span
      className="MuiSlider-track WithStyles(ForwardRef(Slider))-track-3"
      style={
        Object {
          "left": "0%",
          "width": "NaN%",
        }
      }
    />
    <input
      type="hidden"
      value="NaN"
    />
    <span
      aria-orientation="horizontal"
      aria-valuemax={4000}
      aria-valuemin={1200}
      aria-valuenow={NaN}
      className="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-2 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "NaN%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="greyBoldText leftTopMarginStyling"
  >
    MONTHLY INVESTMENT
  </div>
  <div
    className="blackText"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    $
    0
  </div>
  <span
    className="MuiSlider-root WithStyles(ForwardRef(Slider))-root-1 MuiSlider-colorPrimary slider"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-4"
    />
    <span
      className="MuiSlider-track WithStyles(ForwardRef(Slider))-track-3"
      style={
        Object {
          "left": "0%",
          "width": "NaN%",
        }
      }
    />
    <input
      type="hidden"
      value="NaN"
    />
    <span
      aria-orientation="horizontal"
      aria-valuemax={800}
      aria-valuemin={0}
      aria-valuenow={NaN}
      className="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-2 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "NaN%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="greyBoldText leftTopMarginStyling"
  >
    YEARS WITH UP&UP
  </div>
  <div
    className="blackText"
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    1
     
    year
  </div>
  <span
    className="MuiSlider-root WithStyles(ForwardRef(Slider))-root-1 MuiSlider-colorPrimary slider"
    onMouseDown={[Function]}
  >
    <span
      className="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-4"
    />
    <span
      className="MuiSlider-track WithStyles(ForwardRef(Slider))-track-3"
      style={
        Object {
          "left": "0%",
          "width": "NaN%",
        }
      }
    />
    <input
      type="hidden"
      value="NaN"
    />
    <span
      aria-orientation="horizontal"
      aria-valuemax={5}
      aria-valuemin={1}
      aria-valuenow={NaN}
      className="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-2 MuiSlider-thumbColorPrimary"
      data-index={0}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      role="slider"
      style={
        Object {
          "left": "NaN%",
        }
      }
      tabIndex={0}
    />
  </span>
  <div
    className="greyBoldText leftTopMarginStyling"
  >
    OWNERSHIP VALUE
  </div>
  <div
    className="blueTextLarge"
    style={
      Object {
        "alignSelf": "center",
        "marginBottom": "7px",
        "marginTop": "10px",
      }
    }
  >
    $
    0
  </div>
  <div
    className="greyText centerTextStyling"
  >
    Use your $
    0
     ownership stake as a down payment on any home, transfer it to your next up&up rental, or cash out.
  </div>
</div>
`;
